docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc

```diff
@@ -1078,6 +1079,122 @@ public @interface CanWrite {
+Let's consider following Jakarta REST resource:
+
+[source,java]
+----
+package org.acme.security.rest.resource;
+
+import io.quarkus.security.PermissionsAllowed;
+import jakarta.ws.rs.BeanParam;
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+
+@Path("/hello")
+public class SimpleResource {
+
+    @PermissionsAllowed(value = "read", permission = BeanParamPermission.class,
+        params = "beanParam.securityContext.userPrincipal.name") <1>
+    @GET
+    public String sayHello(@BeanParam SimpleBeanParam beanParam) {
+        return "Hello from " + beanParam.uriInfo.getPath();
+    }
+
+}
+----
+<1> The `params` annotation attribute specifies that user principal should be passed to the `BeanParamPermission` constructor.
+Other `BeanParamPermission` constructor parameters like `customAuthorizationHeader` and `query` are matched automatically.
+Quarkus identifies the `BeanParamPermission` constructor parameters among `beanParam` fields and their public accessors.
+To avoid ambiguous resolution, automatic detection only works for the `beanParam` fields.
+For that reason, we had to specify path to the user principal explicitly.
+
+Where the `SimpleBeanParam` class is declared like in the example below:
+
+[source,java]
+----
+package org.acme.security.rest.dto;
+
+import java.util.List;
+
+import jakarta.ws.rs.HeaderParam;
+import jakarta.ws.rs.QueryParam;
+import jakarta.ws.rs.core.Context;
+import jakarta.ws.rs.core.SecurityContext;
+import jakarta.ws.rs.core.UriInfo;
+
+public class SimpleBeanParam {
+
+    @HeaderParam("CustomAuthorization")
+    private String customAuthorizationHeader;
+
+    @Context
+    SecurityContext securityContext;
+
+    @Context
+    public UriInfo uriInfo;
+
+    @QueryParam("query")
+    public String query;    <1>
+
+    public SecurityContext getSecurityContext() { <2>
+        return securityContext;
+    }
+
+    public String customAuthorizationHeader() { <3>
+        return customAuthorizationHeader;
+    }
+}
+----
+<1> Quarkus Security can only pass public fields to a custom permission constructor.
+<2> Quarkus Security automatically uses public getter methods if they are available.
+<3> The `customAuthorizationHeader` field is not public, therefore Quarkus access this field with the `customAuthorizationHeader` accessor.
+That is particularly useful with Java records, where generated accessors are not prefixed with `get`.
+
+Here is an example of the `BeanParamPermission` permission that checks user principal, custom header and query parameter:
+
+[source,java]
+----
+package org.acme.security.permission;
+
+import java.security.Permission;
+
+public class BeanParamPermission extends Permission {
+
+    private final String permissionName;
+    private final String customAuthorization;
+    private final String userName;
+    private final String queryParam;
+
+    public BeanParamPermission(String permissionName, String customAuthorizationHeader, String name, String query) {
```

Comment:
I'm a little bit confused. Oh, I see, you have unwrapped the user name from the security context.
Hmm, it is kind of nice and neat, but very complex at the same time. I'm not sure we need to support such combinations, the security identity for example should be accessible directly from the permission class

Reply:
I doubt typing a string `beanParam.securityContext.userPrincipal.name` is better than using Arc to get SecurityIdentity (as opposed to JAX-RS security context, which we recommend to bind to SecurityIdentity), it feels like we can pass on this one, for now at least

Reply:
> I'm not sure we need to support such combinations, the security identity for example should be accessible directly from the permission class

you can do that, but some people prefer JAX-RS specific concepts like here

Reply:
> I doubt typing a string beanParam.securityContext.userPrincipal.name is better than using Arc to get SecurityIdentity (as opposed to JAX-RS security context, which we recommend to bind to SecurityIdentity), it feels like we can pass on this one, for now at least

I'll remove it from docs example. Please comment if you want more.

Reply:
@sberyozkin will need bit more input here, do you have alternative proposals? I can provide replace this `beanParam.nestedBeanParam.query` but I think you will have same argument. It's legit concern, but we shouldn't assume that users prefer your way. So let's suggest something and I'll do it.

Reply:
Hi Michal, @michalvavrik, it is just that on one hand we don't like EL (mainly due to the untyped nature of it) but then we allow long untyped sequences after all here. But indeed let's keep it and we may end up coming up with some micro EL support in the end after all :-)